### PR TITLE
Add support for caching fetched relations with Criteria

### PR DESCRIPTION
**Pull request by bahusoid**
Follow up to #1952
As for Criteria it's much simpler to collect entities that are already present in `ResultTypes` (and so in cache) I've reversed original implementation - and instead of additional entities to cache provide entities that already cached. And additional entities (fetched entities and collections) are calculated automatically.

### CODE REVIEW

File: src/NHibernate.DomainModel/Northwind/Entities/Northwind.cs [hunk 1]
```diff
@@ -94,9 +94,9 @@ public IQueryable<Role> Role
             get { return _session.Query<Role>(); }
         }
 
-        public IEnumerable<IUser> IUsers
+        public IQueryable<IUser> IUsers
```
Comment (bahusoid) [inline src/NHibernate.DomainModel/Northwind/Entities/Northwind.cs]: Small unrelated fix

File: src/NHibernate/Hql/Ast/ANTLR/Tree/SelectClause.cs [hunk 2]
```diff
@@ -55,7 +55,6 @@ public void InitializeDerivedSelectClause(FromClause fromClause)
 
 			ASTAppender appender = new ASTAppender(ASTFactory, this);	// Get ready to start adding nodes.
 			int size = fromElements.Count;
-			List<IType> sqlResultTypeList = new List<IType>(size);
```
Comment (bahusoid) [inline src/NHibernate/Hql/Ast/ANTLR/Tree/SelectClause.cs]: Small unrelated clean up

**Comment by bahusoid** [comment]
Failing NH2673 (CanCacheCriteriaWithLeftJoin and others) shows subtle difference how criteria and hql is processed.
For hql `QueryLoader.GetResultColumnOrRow` doesn't apply supplied transformer. Transformation is delayed till `GetResultList` call.

In criteria transformer is applied right inside `CriteriaLoader.GetResultColumnOrRow`. And transformer result might change objects type an count. So `Loader.ResultTypes` basically means nothing in criteria when `QueryCacheResultBuilder` tries to cache additional data.

Need to think how to proceed further.  It seems proper fix requires always untransformed results to be sent to cache.

**Comment by bahusoid** [comment]
Nope. I'm wrong. Transformers are applied in both cases before caching. All we need is hql query with fetches and additional data in output and transformer that modifies number of objects in results. So something like this fails in hql on master too:
```C#
var list = session.CreateQuery("select o, o.Customer.CompanyName from Order o join fetch o.Customer")
         .SetCacheable(true)
         .SetResultTransformer(new RootEntityResultTransformer())
         .List<object>();
```

**Comment by bahusoid** [comment]
So the issue: `QueryCacheResultBuilder` builds results based on `loader.ResultTypes`. It expects count to match provided tuple.  But user transformer can modify tuple size  (`RootEntityResultTransformer` is one example) and that breaks `QueryCacheResultBuilder.AddRow` logic.

The fix: put in cache untransformed results for queries that recognized as "queries with fetched relations"

**Comment by bahusoid** [comment]
Found another cache quirk. Entities returned in projections (like `Select(x => x.Customer)`) are not cached. See added test [ProjectedEntitiesAreCachable](https://github.com/nhibernate/nhibernate-core/pull/2090/files#diff-c912479519a3fa87d2676c10172c4ad4R403)

Though this issue is unrelated to changes here or in #1952. As this query doesn't have fetches and processed by old logic.

**Comment by bahusoid** [comment]
Oh common... It seems i'm enumerating here all the possible ways that doesn't work... Oracle can work with duplicated aliases... Except when limits/paging is applied (as for Oracle original query needs to be wrapped in  subquery `select [listOfAliases] from ( originalQuery) where numRow...)` and duplicated aliases with subquery doesn't work)

File: src/NHibernate/Loader/Loader.cs [hunk 3]
```diff
@@ -156,11 +168,18 @@ public virtual bool IsSubselectLoadingEnabled
 		/// </summary>
 		public IType[] ResultTypes { get; protected set; }
 
-		public bool[] EntityFetches { get; protected set; }
+		public IType[] CacheTypes => CacheInfo?.CacheTypes ?? ResultTypes;
```
[the remark this answers is not in the document]
Comment (bahusoid) [reply src/NHibernate/Loader/Loader.cs:171]: Regarding all binary changes comments: I will be able to check next week but I'm pretty sure it's all about stuff that was added in https://github.com/nhibernate/nhibernate-core/pull/1952

So it's not a breaking change as those changes are not released yet.

[the remark this answers is not in the document]
Comment (bahusoid) [reply src/NHibernate/Loader/Loader.cs:171]: https://github.com/nhibernate/nhibernate-core/pull/1952/files#diff-cb653c94f246990d1cf339129e010df9R159-R164

File: src/NHibernate.Test/Criteria/ReadonlyTests/QueryOverCacheableTests.cs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+using NHibernate.SqlCommand;
+using NUnit.Framework;
+
+namespace NHibernate.Test.Criteria.ReadonlyTests
+{
+	[TestFixture]
+	public class QueryOverCacheableTests : CriteriaNorthwindReadonlyTestCase
+	{
+		//Just for discoverability
+		private class CriteriaCacheableTest{}
```
[the remark this answers is not in the document]
Comment (bahusoid) [reply src/NHibernate.Test/Criteria/ReadonlyTests/QueryOverCacheableTests.cs:13]: To make it appear in search when user enters Criteria trying to find some Criteria related tests

**Comment by bahusoid** [comment]
And why bug? IMHO it's a new feature

File: src/NHibernate.Test/Criteria/ReadonlyTests/QueryOverCacheableTests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (fredericDelaporte) [reply src/NHibernate.Test/Criteria/ReadonlyTests/QueryOverCacheableTests.cs:13]: Can-it be just a comment, not a dummy class?

**Comment by fredericDelaporte** [comment]
I have classified it as bug for consistency with #1195 which was fixed by #1952. It was considered there that not caching fetches in Linq was a bug, so why not doing the same for criteria/query-over? Otherwise we should re-qualify #1195.

File: src/NHibernate/Transform/CacheableResultTransformer.cs [hunk 5]
```diff
@@ -60,14 +61,25 @@ public class CacheableResultTransformer : IResultTransformer
...
+		[Obsolete("Please use overload with skipTransformer parameter.")]
 		public static CacheableResultTransformer Create(IResultTransformer transformer,
 		                                                string[] aliases,
 		                                                bool[] includeInTuple)
 		{
 			return Create(transformer, aliases, includeInTuple, false, null);
 		}
 
+		// Since 5.2
+		[Obsolete("Please use override with skipTransformer parameter.")]
```
Comment (bahusoid) [inline src/NHibernate/Transform/CacheableResultTransformer.cs:73]:
```suggestion
		[Obsolete("Please use overload with skipTransformer parameter.")]
```

File: src/NHibernate/Transform/CacheableResultTransformer.cs [hunk 6]
```diff
@@ -60,14 +61,25 @@ public class CacheableResultTransformer : IResultTransformer
 		/// <returns>a CacheableResultTransformer that is used to transform
 		///    tuples to a value(s) that can be cached.</returns>
 		// Since v5.1
-		[Obsolete("Please use overload with autoDiscoverTypes parameter.")]
+		[Obsolete("Please use override with skipTransformer parameter.")]
```
[the remark this answers is not in the document]
Comment (bahusoid) [reply src/NHibernate/Transform/CacheableResultTransformer.cs]: I didn't change wording by intention. I wrote comment for other method and just copy/pasted without much thinking. Fixed other case too

File: src/NHibernate.Test/Criteria/ReadonlyTests/QueryOverCacheableTests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bahusoid) [reply src/NHibernate.Test/Criteria/ReadonlyTests/QueryOverCacheableTests.cs:13]: >Can-it be just a comment, not a dummy class?

When I'm not sure which test fixture I need I often try to find fixture class I need by some keywords. So comment is less discoverable than dummy class. Well if you really dislike it  - you can drop it. 

>  It is there just because query-over is an added layer to criteria and so test it by the way, isn't-it?

Yes.

**Comment by bahusoid** [comment]
> It was considered there that not caching fetches in Linq was a bug, so why not doing the same for criteria/query-over

Hm... Ok. I didn't pay attention how it was classified for Linq/hql. But if re-qualifying means only changing labels - let's re-qualify. IMO those PRs should be stated in release in New Features list.
